File: MessageServer/JavaEE/MessageServer/src/Managers/sync/GroupManager.java
package Managers.sync;

import java.util.HashSet;
import java.util.List;

import DBConnection.DBMessage;
import DBConnection.DBResult;
import Entities.Message;
import Entities.MsgType;
import Entities.User;
import Managers.Manager;
import Websocket.ServerManager;

public class GroupManager implements Manager {

	HashSet<String> members;
	HashSet<String> invitees;
	String creator;
	String groupName;

	public GroupManager(String creator, String groupName) {
		this.creator = creator;
		this.groupName = groupName;
		members = new HashSet<>();
		members.add(creator);
		invitees = new HashSet<>();
	}

	public GroupManager(String creator, String groupName, List<String> member, List<String> invitee) {
		this.creator = creator;
		this.groupName = groupName;
		members = new HashSet<>();
		for (String entry : member) {
			members.add(entry);
		}
		members.add(creator);
		invitees = new HashSet<>();
		for (String entry : invitee) {
			invitees.add(entry);
		}
	}

	@Override
	public void handleMsg(Message message) {
		//System.out.println("GroupManager " + groupName + "Received Msg " + message.toString());
		if (message != null && (invitees.contains(message.senderID) || members.contains(message.senderID))) {
			switch (message.type) {
			case Private:
				sendMessage(message);
				break;
			case AcceptRequest:
				acceptRequest(message);
				break;
			case DeclineRequest:
				declineRequest(message);
				break;
			case LeaveGroup:
				leaveGroup(message);
				break;
			case InviteToGroup:
				inviteToGroup(message);
				break;
			default:
				break;
			}
		}
	}

	private void acceptRequest(Message message) {
		//System.out.println("GroupManager: User" + message.senderID + " accepted request");
		if (invitees.contains(message.senderID)) {
			members.add(message.senderID);
			invitees.remove(message.senderID);
			boolean result = DBMessage.updateGroupEntry(this.groupName, message.senderID, true, 3);
			if (!result) {
				ServerManager.sendBackErrorMsg(message);
			}

		}
	}

	private void declineRequest(Message message) {
		invitees.remove(message.senderID);
		boolean result = DBMessage.removeGroupEntry(this.groupName, message.senderID, 3); 
		if (!result) {
			ServerManager.sendBackErrorMsg(message);
		}
	}

	private void leaveGroup(Message message) {
		members.remove(message.senderID);
		boolean result = DBMessage.removeGroupEntry(this.groupName, message.senderID, 3); 
		if (!result) {
			ServerManager.sendBackErrorMsg(message);
		}
	}

	private void inviteToGroup(Message message) {
		if (!(members.contains(message.content) && invitees.contains(message.content))) {
			DBResult<User> dbUser = DBMessage.getUser(message.content, 3);
			User user = dbUser.result;
			if (dbUser.success) {
				if (user != null) {
					invitees.add(message.content);					
					boolean result = DBMessage.updateGroupEntry(this.groupName, user.id, false, 3);
					if (result) {
						ServerManager.handleMsg(new Message(groupName, message.content, "", MsgType.InviteToGroup));
					} else {
						ServerManager.sendBackErrorMsg(message);
					}
				}
			} else {
				ServerManager.sendBackErrorMsg(message);
			}
		}
	}

	private void sendMessage(Message message) {
		Message send;
		//System.out.println("GroupManager : " + groupName + "sends message to members");
		for (String member : members) {
			send = new Message(groupName + " -> " + message.senderID, member, message.content, message.type);
			ServerManager.handleMsg(send);
		}
	}

	public String getCreator() {
		return this.creator;
	}

}

File: MessageServer/RxKotlin/MessageServer/src/main/kotlin/at/mh/kotlin/message/server/manager/GroupManager.kt
package at.mh.kotlin.message.server.manager

import at.mh.kotlin.message.server.db.DBConnR2DBC
import at.mh.kotlin.message.server.messages.*
import io.reactivex.Observable
import io.reactivex.disposables.Disposable
import io.reactivex.rxjava3.core.Observer
import org.apache.logging.log4j.LogManager
import java.util.concurrent.TimeUnit

/**
 * Handles all the messages of a Group
 */
class GroupManager(
    private val groupName: String,
    val groupCreator: String,
    private val members: HashSet<String>,
    private val invitees: HashSet<String>,
    private val db: DBConnR2DBC,
    private val serverManager: ServerManager
) : Manager() {
    private val logger = LogManager.getLogger("GroupManager")
    private var serverMgr: Observer<in Message>? = null
    private var disposable: Disposable? = null

    /**
     * can be used to remove Groups if they haven't received messages for a specified amount of time
     */
    private fun startTimeout(timeInMinutes: Long) {
        disposable?.dispose()
        disposable = Observable.timer(timeInMinutes, TimeUnit.MINUTES).doOnNext {
            logger.error("$groupName hasn't received messages since $timeInMinutes min removing him from active list")
            this.serverMgr?.onComplete()
            serverManager.removeManager(groupName)
        }.subscribe()
    }

    override fun onComplete() {
    }

    /**
     * Handles all an incoming Message
     */
    override fun onNext(message: Message) {
        logger.info("$groupName received message $message")
        when (message) {
            is PrivateMessage -> sendMessageToMembers(message)
            is InviteToGroup -> inviteToGroup(message)
            is AcceptRequest -> addToGroup(message)
            is DeclineRequest -> removeFromInvitees(message)
            is LeaveGroup -> leaveGroup(message)
            else -> logger.error("Unrecognized Message $message")
        }
    }

    override fun onError(e: Throwable?) {
        logger.error("$groupName received onError ${e?.message}")
    }


    override fun subscribeActual(manager: Observer<in Message>?) {
        this.serverMgr = manager
    }

    /**
     * Sends the received message to all of its members
     */
    private fun sendMessageToMembers(message: PrivateMessage) {
        logger.info(" $groupName Private message received $message ${members.firstOrNull()}")
        members.forEach { serverMgr?.onNext(PrivateMessage(groupName, it, message.content)) }
    }

    /**
     * Initiate an invite
     */
    private fun inviteToGroup(message: InviteToGroup) {
        // check if initiator is a group member
        if (members.contains(message.senderID)) {
            logger.info("$groupName received InviteRequest")
            // check if invitee isn't already a member and hasn't already received an invite
            if (!members.contains(message.content) && !invitees.contains(message.content)) {
                // add user to invitees and update GroupEntry in DB
                db.getUser(message.content)
                    .onErrorComplete {
                        logger.info("DB-Error $it")
                        true
                    }
                    .subscribe { userID ->
                        logger.info("User found $userID")
                        db.updateGroupEntry(groupName, message.content, accepted = false)
                            .onErrorComplete {
                                logger.error("DB-Error $it")
                                false
                            }
                            .doOnComplete {
                                logger.info("Invite saved in to DB")
                                invitees.add(message.content)
                                serverMgr?.onNext(InviteToGroup(groupName, message.content, message.content))
                            }.subscribe()

                    }
            } else {
                logger.info("${message.content} already in invitees or group")
            }
        } else {
            logger.info("sender is not in members ${message.senderID}")
        }

    }

    /**
     * Add a User to the Group
     */
    private fun addToGroup(message: AcceptRequest) {
        logger.info("user accepted request $message")
        // Check if AcceptMsg was sent by an invitee
        if (invitees.contains(message.senderID) && !members.contains(message.senderID)) {
            db.updateGroupEntry(groupName, message.senderID, true)
                .onErrorComplete {
                    logger.warn("DB-Error $it")
                    false
                }
                .doOnComplete {
                    logger.info("User added to Members in DB")
                    // add to members and update GroupEntry
                    members.add(message.senderID)
                    invitees.remove(message.senderID)
                }.subscribe()
        }
    }

    /**
     * Invitee declined request
     */
    private fun removeFromInvitees(message: DeclineRequest) {
        // check if sender was an invitee and then remove them
        if (invitees.contains(message.senderID)) {
            db.removeGroupEntry(groupName, message.senderID)
                .onErrorComplete {
                    logger.warn("DB-Error $it")
                    false
                }
                .doOnComplete {
                    logger.info("Invitee removed ${message.senderID}")
                    invitees.remove(message.senderID)
                }.subscribe()
        }
    }

    /**
     * Remove member from group
     */
    private fun leaveGroup(message: LeaveGroup) {
        if (members.contains(message.senderID)) {
            db.removeGroupEntry(groupName, message.senderID)
                .onErrorComplete {
                    logger.warn("DB-Error $it")
                    false
                }.doOnComplete {
                    logger.info("User left group successfully ${message.senderID}")
                    members.remove(message.senderID)
                }
        }
    }
}